Documentation for entitlements (#152)
onflow/docs: docs/cadence/language/references.mdx

`auth(E, F, G) &T` for a reference authorized to `E`, `F` and `G`,
or `auth(E | F | G) &T` for a refernece authorized to `E`, `F`, **or** `G`. 
Authorized references are subtypes of unauthorized references.

Possessing an entitlement allows a reference to have access to any functions on its referenced type
that require that entitlement. E.g, if we extended the `HasCount` interface with a function:

```cadence
entitlement Reset

resource interface HasCount {
    count: Int
    access(Reset) fun resetCount() 
}
```

Then an unauthorized reference of type `&{HasCount}` would be unable to call `resetCount`. 
However, we can create a reference that can, like so:

```
let authCountRef: auth(Reset) &{HasCount} = &counter

// Valid, because `authCountRef` is authorized to `Reset`
authCountRef.resetCount()
```

It is important to note that while references are covariant (and downcastable) with respect to their reference type, 
the authorization portion of the reference can never be downcast. 
In fact, the only way to "add" entitlements to a reference is to do so at the time of its creation, like in the example above. 

> An reference will never have any more entitlements than the set with which it was created, 

Reviewer: ```suggestion
A reference will never have any more entitlements than the set with which it was created, 
```

//
countRef.increment()

// We can conditionally downcast `countRef` to a `Counter` if it has
// that type at runtime. 
//
let counterRef2: &Counter = countRef as? &Counter
counterRef2.increment()
```

References are ephemeral, i.e. they cannot be [stored](accounts#account-storage).
Instead, consider [storing a capability and borrowing it](capability-based-access-control) when needed.

## Authorized References

By default, references are **unauthorized**. 

> However, they may also be **authorized** to a set of [entitlements](./access-control.mdx#entitlements)

Reviewer 2: ```suggestion
However, they may also be **authorized** to a set of [entitlements](./access-control.md#entitlements)
```

Authorized references have the `auth` modifier, 
along with the set of entitlements to which they are authorized. The full syntax is:
`auth(E, F, G) &T` for a reference authorized to `E`, `F` and `G`,
or `auth(E | F | G) &T` for a refernece authorized to `E`, `F`, **or** `G`. 
Authorized references are subtypes of unauthorized references.

Entitlements can only be given to references when they are created, 
and references to a value can only be created by the owner of the value. 
When creating a reference, that reference can be given any set of entitlements the value owner wishes to add. 

> Possessing an entitlement allows a reference to have access to any functions on its referenced type

Reviewer 3: ```suggestion
Possessing an entitlement allows a reference to have access to functions and fields on its referenced type
```

then `{Us}` is a subset of `{Vs}` when any of the `Us` also appears in `{Vs}`. 
To see why, consider again that `{Vs}` expresses a set of entitlements that the reference **might** possess, 
and as long as at least one of these entitlements is in `{Us}` (which is a set of entitlements that we **know** the reference has), 
then the description provided by `{Vs}` is correct. 

As an example to illustrate these rules:

```
let eRef: auth(E) &T = ...
let efRef: auth(E, F) &T = ...
let eOrFRef: auth(E | F) &T = ...

// Invalid, `eRef` only has `E` but `F` is required
eRef as auth(F) &T

// Invalid, `eRef` only has `E` but both `E` and `F` are required
eRef as auth(E, F) &T

// Valid, `eRef` definitely has `E` and either `E` or `F` is sufficient
eRef as auth(E | F) &T

// Valid, `efRef` both `E` and `F` but only `F` is required
efRef as auth(F) &T

// Valid, `efRef` both `E` and `F`, and either is sufficient
efRef as auth(E | F) &T

// Invalid, `eOrFRef` has one of `E` or `F` but we need to definitely have `F`
eOrFRef as auth(F) &T

// Invalid, `eOrFRef` has one of `E` or `F` but we need both
eOrFRef as auth(E, F) &T
```

### References and Entitlement Mappings

> In most situations, an [entitlement mapping](./access-control.mdx#entitlement-mappings) is valid in the `auth` portion of a reference type.

Reviewer 2: ```suggestion
In most situations, an [entitlement mapping](./access-control.md#entitlement-mappings) is valid in the `auth` portion of a reference type.
```
